akka-remote/src/main/scala/akka/remote/artery/TcpFraming.scala

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+import scala.annotation.tailrec
+
+import akka.annotation.InternalApi
+import akka.stream.Attributes
+import akka.util.ByteString
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.stream.stage.StageLogging
+import akka.remote.artery.FlightRecorderEvents.TcpInbound_Received
+import akka.stream.scaladsl.Framing.FramingException
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object TcpFraming {
+  val Undefined = Int.MinValue
+
+  def frameHeader(size: Int) =
+    ByteString(
+      (size & 0xff).toByte,
+      ((size & 0xff00) >> 8).toByte,
+      ((size & 0xff0000) >> 16).toByte,
+      ((size & 0xff000000) >> 24).toByte
+    )
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class TcpFraming(flightRecorder: EventSink) extends GraphStage[FlowShape[ByteString, EnvelopeBuffer]] {
```

Comment:
Here's a simple implementation based on `ByteStringParser`:

```scala
import akka.stream.impl.io.ByteStringParser
import akka.stream.impl.io.ByteStringParser.ByteReader
import akka.stream.impl.io.ByteStringParser.ParseResult
import akka.stream.impl.io.ByteStringParser.ParseStep

/**
 * INTERNAL API
 */
@InternalApi private[akka] class TcpFraming(flightRecorder: EventSink) extends ByteStringParser[EnvelopeBuffer] {
  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new ParsingLogic {
    abstract class Step extends ParseStep[EnvelopeBuffer]
    startWith(ReadStreamId)

    case object ReadStreamId extends Step {
      override def parse(reader: ByteReader): ParseResult[EnvelopeBuffer] =
        ParseResult(None, ReadFrame(reader.readByte()))
    }
    case class ReadFrame(streamId: Int) extends Step {
      override def onTruncation(): Unit =
        failStage(new FramingException("Stream finished but there was a truncated final frame in the buffer"))

      override def parse(reader: ByteReader): ParseResult[EnvelopeBuffer] = {
        val frameLength = reader.readIntLE()
        val buffer = createBuffer(reader.take(frameLength))
        ParseResult(Some(buffer), this)
      }

      private def createBuffer(bs: ByteString): EnvelopeBuffer = {
        val buffer = ByteBuffer.wrap(bs.toArray)
        buffer.order(ByteOrder.LITTLE_ENDIAN)
        flightRecorder.hiFreq(TcpInbound_Received, buffer.limit)
        val res = new EnvelopeBuffer(buffer)
        res.setStreamId(streamId)
        res
      }
    }
  }
}
```

Good thing is that you don't need to care yourself for all the push/pull handling and restarting parsing when too little data is available.

Reply:
thanks, I'll switch to that